### ISSUE TITLE
auto-wrap tensors as inputs to autograd

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#274]
A request from @jcjohnson that's pretty reasonable, needs to be implemented

> Having to feed Variables as input to Modules is somewhat annoying, especially since DataLoader provides Tensors. On the one hand explicitly constructing variables makes it clear that Tensors and Variables are different beasts, but on the other hand it is a bit of syntactic noise that shows up in a lot of places. Would it be a terrible idea to have the forward method of a Module autowrap input Tensors into Variables?

**Comment by soumith** [comment]
i'm going to close this issue. There hasn't been any movement in discussion, and I presume folks have developed their own solutions for this. The choice between making an input `Tensor` to be `requires_grad=False` or `volatile=True` is what we have failed to reach consensus on.

**Comment by adamlerer** [comment]
Lets implement this. 

I think that tensors should definitely be converted to Variable with `requires_grad=False`, *not* `volatile=True`, because (a) the fact that you provide one input as a tensor in no way implies that you don't need gradients of other inputs; (b) `volatile` is just an optimization; better to leave it functional by default, and the user can always set `volatile=True` (or disable gradients another way like `model.eval()`) if they need that optimization.

**Comment by jcjohnson** [comment]
I agree with `requires_grad=False`, `volatile=False` for automatic Tensor to Variable conversion for the same reasons @adamlerer mentions.

I think the main concern is the issue raised by @jekbradbury: Some users won't realize that this conversion is happening, and it will be pretty unobvious that they should manually construct a Variable with `volatile=True` if they want to run inference with more efficient memory usage.

**Comment by apaszke** [comment]
Yes, of course. I think `requires_grad=True` is the only choice that makes sense.

**Comment by jekbradbury** [comment]
Yes, I agree. BTW the Chainer devs are moving to the same API (they’ve always autowrapped Tensors but now they’re also adding a requires_grad attribute to Variable with the same semantics as PyTorch, defaulting to False, so that they don’t have to have to allow modules to return both tensors and variables).

**Comment by fmassa** [comment]
Tensors and Variables are the same thing now.
